src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java

```diff
@@ -0,0 +1,183 @@
+import static com.mysql.cj.util.StringUtils.isNullOrEmpty;
+
+import com.mysql.cj.conf.ConnectionUrl;
+import com.mysql.cj.conf.PropertySet;
+import com.mysql.cj.log.Log;
+import com.mysql.cj.util.LRUCache;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;
+import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueResponse;
+import software.amazon.awssdk.services.secretsmanager.model.ResourceNotFoundException;
+
+import java.sql.SQLException;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.concurrent.Callable;
+
+public class AWSSecretsManagerPlugin implements IConnectionPlugin {
+  private static final String SECRET_ID_PROPERTY = "secretsManagerSecretId";
+  private static final String REGION_PROPERTY = "secretsManagerRegion";
+  private static final String ERROR_MISSING_DEPENDENCY =
+      "[AWSSecretsManagerPlugin] Required dependency 'AWS Java SDK for AWS Secrets Manager' is not on the classpath";
+  private static final String ERROR_GET_SECRETS_FAILED =
+      "[AWSSecretsManagerPlugin] Was not able to either fetch or read the database credentials from AWS Secrets Manager. Ensure the correct secretId and region properties have been provided";
+  private static final String ERROR_INITIAL_CONN_FAILED =
+      "[AWSSecretsManagerPlugin] Open initial connection failed with given credentials. Retrieving secrets was not attempted";
+  private static final String SQLSTATE_ACCESS_ERROR = "28000";
+
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+  private final String secretId;
+  private final Region region;
+  private final LRUCache<String, String> secretCache = new LRUCache<>(100);
+
+  public AWSSecretsManagerPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) throws SQLException {
+
+    try {
+      Class.forName("software.amazon.awssdk.services.secretsmanager.SecretsManagerClient");
+    } catch (ClassNotFoundException e) {
+      logger.logError(ERROR_MISSING_DEPENDENCY);
+      throw new SQLException(ERROR_MISSING_DEPENDENCY);
+    }
+
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+    this.secretId = propertySet.getStringProperty(SECRET_ID_PROPERTY).getValue();
+    this.region = Region.of(propertySet.getStringProperty(REGION_PROPERTY).getValue());
+  }
+
+  @Override
+  public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
+    final Properties updatedProperties = new Properties();
+    updatedProperties.putAll(connectionUrl.getOriginalProperties());
+    String secret = secretCache.get(this.secretId);
```

Comment:
Is secretId unique among all regions? Or secretId is unique within a single region? A key for LRU cache may need to include region as well.

Reply:
How we handle a case where cache is empty and secret is null?

Reply:
The cache stores no a parsed user name and password but a JSON string that contains them. That means that every time the plugin needs user name and password, it needs to parse JSON. It's a bit of wasting a time. Is it possible to store pre-parsed value in the cache?

Reply:
If the cache is empty, the driver will attempt to get the secret. If there are any exceptions when getting that secret, including if the secret does not exist, the driver will throw an SQLException.